test/test_device_radix_sort.cu

```diff
@@ -982,16 +1002,67 @@ void TestBits(
         int begin_bit = 1;
         int end_bit = (sizeof(KeyT) * 8) - 1;
         printf("Testing key bits [%d,%d)\n", begin_bit, end_bit); fflush(stdout);
-        TestDirection(h_keys, num_items, num_segments, h_segment_offsets, begin_bit, end_bit);
+        TestDirection(h_keys, num_items, num_segments, h_segment_offsets, d_segment_begin_offsets, d_segment_end_offsets, begin_bit, end_bit);
 
         // Across subword boundaries
         int mid_bit = sizeof(KeyT) * 4;
         printf("Testing key bits [%d,%d)\n", mid_bit - 1, mid_bit + 1); fflush(stdout);
-        TestDirection(h_keys, num_items, num_segments, h_segment_offsets, mid_bit - 1, mid_bit + 1);
+        TestDirection(h_keys, num_items, num_segments, h_segment_offsets, d_segment_begin_offsets, d_segment_end_offsets, mid_bit - 1, mid_bit + 1);
     }
 
     printf("Testing key bits [%d,%d)\n", 0, int(sizeof(KeyT)) * 8); fflush(stdout);
-    TestDirection(h_keys, num_items, num_segments, h_segment_offsets, 0, sizeof(KeyT) * 8);
+    TestDirection(h_keys, num_items, num_segments, h_segment_offsets, d_segment_begin_offsets, d_segment_end_offsets, 0, sizeof(KeyT) * 8);
+}
+
+
+template<typename OffsetT>
+struct transform_functor1
+{
+    __host__ __device__ __forceinline__ OffsetT operator()(OffsetT offset) const
+    {
+        return offset;
+    }
+};
+
+template<typename OffsetT>
+struct transform_functor2
+{
+    __host__ __device__ __forceinline__ OffsetT operator()(OffsetT offset) const
+    {
+        return offset;
+    }
+};
+
+
+/**
+* Test different segment iterators
+*/
+template <typename KeyT>
+void TestSegmentIterators(
+    KeyT    *h_keys,
+    int     num_items,
+    int     num_segments,
+    int     *h_segment_offsets,
+    int     *d_segment_offsets)
+{
+    InitializeSegments(num_items, num_segments, h_segment_offsets);
+    CubDebugExit(cudaMemcpy(d_segment_offsets, h_segment_offsets, sizeof(int) * (num_segments + 1), cudaMemcpyHostToDevice));
+
+    // Test with segment pointer
+    TestBits(h_keys, num_items, num_segments, h_segment_offsets, d_segment_offsets, d_segment_offsets + 1);
+
+    // Test with segment iterator
+    typedef CastOp<int> IdentityOpT;
+    IdentityOpT identity_op;
+    TransformInputIterator<int, IdentityOpT, int*, int> d_segment_offsets_itr(d_segment_offsets, identity_op);
+
+    TestBits(h_keys, num_items, num_segments, h_segment_offsets, d_segment_offsets_itr, d_segment_offsets_itr + 1);
+
+    // Test with thrust transform iterators
+    auto d_segment_begin_offsets_itr = thrust::make_transform_iterator(d_segment_offsets, transform_functor1<int>());
```

Comment:
This adds an include for `cub/iterator/transform_input_iterator.cuh`, but uses Thust's transform iterators. Please update this to use the CUB iterators in the implementation.

Reply:
I used Thrust's iterators because they were mentioned in the problem description. Are you completely sure the Thrust's iterators should be replaced with the CUB ones?

Reply:
Yes -- as long as the offset iterators are different types, the test will be sufficient for the feature. It doesn't matter which transform iterator implementation is used.

For context, right now there is a circular dependency between Thrust and CUB that I'm trying to break. Thrust must depend on CUB for its algorithm implementations, but CUB doesn't have to depend on Thrust, so I'm trying to avoid adding any new Thrust usages in CUB.